src/workspaces/infrastructure/persistence/repositories/workspace.repository.ts

```diff
@@ -19,4 +24,71 @@ export class WorkspaceRepository {
     );
     return WorkspaceMapper.toDomain(newEntity);
   }
+
+  async findManyWithPagination(
+    userId: User['id'],
+    paginationOptions: IPaginationOptions,
+  ): Promise<Workspace[]> {
+    const entities = await this.workspaceRepository.find({
+      where: {
+        members: {
+          id: userId as number,
+        },
+      },
+      skip: (paginationOptions.page - 1) * paginationOptions.limit,
+      take: paginationOptions.limit,
+    });
+    return entities.map((workspace) => WorkspaceMapper.toDomain(workspace));
+  }
+
+  async findUsersWithPagination(
+    workspaceId: Workspace['id'],
+    paginationOptions: IPaginationOptions,
+  ): Promise<User[]> {
+    const workspace = await this.workspaceRepository.findOne({
+      where: {
+        id: workspaceId as number,
+      },
+      relations: {
+        members: true,
+      },
+    });
+
+    if (!workspace) {
+      throw new Error('Workspace not found');
+    }
+
+    const members = workspace.members.slice(
+      (paginationOptions.page - 1) * paginationOptions.limit,
+      paginationOptions.page * paginationOptions.limit,
+    );
+
+    return members;
+  }
+
+  async findOne(
+    fields: EntityCondition<Workspace>,
+  ): Promise<NullableType<Workspace>> {
+    const entity = await this.workspaceRepository.findOne({
+      where: fields as FindOptionsWhere<WorkspaceEntity>,
+    });
+
+    return entity ? WorkspaceMapper.toDomain(entity) : null;
+  }
+
+  async softDelete(id: Workspace['id'], userId: User['id']): Promise<void> {
+    const entity = await this.workspaceRepository.findOne({
+      where: {
+        id: Number(id),
+        owner: {
+          id: Number(userId),
+        },
+      },
+    });
+    if (!entity) {
+      throw new Error('Workspace not found');
+    }
+
```

Comment:
`softDelete` should only take `Workspace['id']`, and if the user is not the owner must get `unauthorized 401` rather than `Internal server error`, so I think the authorization step can be one step earlier at `worskspace.service`.

With this, changing the authorization technique in the future will only affect the service level, not how the repository functions.
```suggestion
  async softDelete(id: Workspace['id']): Promise<void> {
    await this.workspaceRepository.softDelete(id)
  }
```